src/lib/homestore_backend/hs_blob_manager.cpp

```diff
@@ -1,17 +1,220 @@
 #include "hs_homeobject.hpp"
+#include "replication_message.hpp"
+#include "replication_state_machine.hpp"
+#include "lib/homeobject_impl.hpp"
+
+SISL_LOGGING_DECL(blobmgr)
 
 namespace homeobject {
 
-BlobManager::Result< blob_id_t > HSHomeObject::_put_blob(ShardInfo const&, Blob&&) {
-    return folly::makeUnexpected(BlobError::UNKNOWN);
+BlobManager::Result< blob_id_t > HSHomeObject::_put_blob(ShardInfo const& shard, Blob&& blob) {
+    auto pg_id = shard.placement_group;
+    HS_PG* pg{nullptr};
+    {
+        std::shared_lock lg{_pg_lock};
+        if (auto it = _pg_map.find(pg_id); it != _pg_map.end()) {
+            pg = s_cast< HS_PG* >(it->second.get());
+        } else {
+            LOGWARN("failed to put blob with non-exist pg [{}]", pg_id);
+            return folly::makeUnexpected(BlobError::UNKNOWN_PG);
+        }
+    }
+
+    if (!pg->repl_dev_) {
+        LOGWARN("failed to get repl dev instance for pg [{}]", pg_id);
+        return folly::makeUnexpected(BlobError::PG_NOT_READY);
+    }
+
+    const uint32_t needed_size = sizeof(ReplicationMessageHeader);
+    auto req = repl_result_ctx< BlobManager::Result< BlobInfo > >::make(needed_size, 512);
+
+    uint8_t* raw_ptr = req->hdr_buf_.bytes;
+    ReplicationMessageHeader* header = new (raw_ptr) ReplicationMessageHeader();
+    header->msg_type = ReplicationMessageType::PUT_BLOB_MSG;
+    header->payload_size = 0;
+    header->payload_crc = 0;
+    header->shard_id = shard.id;
+    header->pg_id = pg_id;
+    header->header_crc = header->calculate_crc();
+
+    // TODO add metadata offset and metadata.
+    BlobHeader blob_header{};
+    blob_header.shard_id = shard.id;
+    blob_header.total_size = blob.body.size;
+    blob_header.hash_algorithm = BlobHeader::HashAlgorithm::CRC32;
+    blob_header.compute_blob_hash(blob.body, false /* verify */);
+
+    sisl::sg_list sgs;
+    sgs.size = 0;
+    auto block_size = pg->repl_dev_->get_blk_size();
+    auto size = sisl::round_up(sizeof(blob_header), block_size);
+    auto buf_header = iomanager.iobuf_alloc(block_size, size);
+    std::memcpy(buf_header, &blob_header, sizeof(blob_header));
+
+    // Create blob header.
+    sgs.iovs.emplace_back(iovec{.iov_base = buf_header, .iov_len = size});
+    sgs.size += size;
+
+    // Append blob bytes.
+    sgs.iovs.emplace_back(iovec{.iov_base = blob.body.bytes, .iov_len = blob.body.size});
+    sgs.size += blob.body.size;
+
+    pg->repl_dev_->async_alloc_write(req->hdr_buf_, sisl::blob{}, sgs, req);
+    auto result = req->result().get();
```

Comment:
This should probably read something like:
```
retun req->result().deferValue([this] ...) -> Result<blob_id_t> {
   header->~ReplicationMessage..;
   ...
   return blob_info.blob_id;
  });
}
```
Otherwise this will block on the async call in the current thread which is not what we want (I think?).